pwnlib/libcdb.py

```diff
@@ -100,11 +110,76 @@ def provider_libc_rip(hex_encoded_id, hash_type):
         return None
     return data
 
+# Search offline https://github.com/niklasb/libc-database for symbols
+def find_local_libc_database(params):
+    if not local_database["libc-database"]:
+        log.warn_once("No environment variable LOCAL_LIBC_DATABASE or parameter libcdb.local_database['libc-database'] are set")
+        return None
+    
+    if not params.get("symbols"):
+        return None
+
+    res = []
+    db_path = Path(local_database["libc-database"]) / "db"
+
+    for symbols_path in db_path.rglob("*.symbols"):
+        syms = get_libc_symbols(symbols_path)
+
+        matched = 0
+        for name, addr in params["symbols"].items():
+            if isinstance(addr, str):
+                addr = int(addr, 16) 
+
+            if syms.get(name) and syms.get(name) & 0xfff == addr & 0xfff:
+                matched += 1
+
+        if matched == len(params["symbols"]):
+            res.append(get_libc_info(db_path, symbols_path.stem, syms))
+
+    return res
+
+# Offline search https://github.com/niklasb/libc-database for hash type
+def offline_provider_libc_database(hex_encoded_id, hash_type):
+    if not local_database["libc-database"]:
+        log.warn_once("No environment variable LOCAL_LIBC_DATABASE or parameter libcdb.local_database['libc-database'] are set")
+        return None
+
+    db_path = Path(local_database["libc-database"]) / "db"
+    hashfilehex = None
+
+    if hash_type == "build_id":
+        hash_type = "buildid"
+
+    # "buildid" is not in HASHES, so we can correct calculate file hash
+    if hash_type in HASHES:
+        hashfilehex = eval(f"{hash_type}filehex")
+
+    for libc_path in db_path.rglob("*.so"):
+        if hash_type == "buildid" and hex_encoded_id == get_build_id(libc_path):
+            return read(libc_path)
+        elif hashfilehex is not None and hex_encoded_id == hashfilehex(libc_path):
+            return read(libc_path)
+
+    return None
+
+
 PROVIDERS = [provider_libcdb, provider_libc_rip]
+OFFLINE_PROVIDERS = [offline_provider_libc_database]
 
-def search_by_hash(hex_encoded_id, hash_type='build_id', unstrip=True):
+
+def search_by_hash(hex_encoded_id, hash_type='build_id', unstrip=True, offline=False):
     assert hash_type in HASHES, hash_type
 
+    # Offline search don't use libcdb cache
+    if offline:
```

Comment:
I just wanted to add an offline search feature to libcdb without intending to make extensive code modifications. Additionally, I prefer not to entirely delegate script behavior control to tools, so I still hope to have an `offline` parameter.

The "offline" and "online" modes have different search logics, and adding the `offline` parameter would also make the code writing process more convenient.

>I'd prefer to just search through the local libc-database if it is configured through the environment variable instead of having to pass `offline=True` everywhere

After configuring the environment variables, if the user hasn't explicitly passed the `offline` parameter, we can automatically set it to `True`.

>So just adding the offline provider to the start of the `PROVIDER` list would be the most convenient way to integrate this?

To distinguish them from the online providers, it's important to note that these two types of provider functions have different return values and behave differently.

Reply:
I would also prefer searching the offline db first, if it exists, and then the online db (unless explicitly disabled with some `--offline` switch). There is no harm searching the offline db in the online mode, is there?

Reply:
Perhaps we can add a new parameter to enable this search mode.